### ISSUE TITLE
[plugins] extras plugin to collect user defined cmds output?

### ISSUE THREAD

**Issue by pmoravec** [sosreport/sos#1663]
I have an idea that will help rather downstream maintenance than to sos upstream. The user story behind this follows:

- somebody wants to collect some files or command outputs
- upstream PR raised, merged after some time
- backported to downstream, released after another bunch of time

The whole meantime, `sosreport` does not collect the required data.

One way how to minimize this time lapse and to be more flexible is to have an `extras` plugin that will:
- open a predefined file (`/etc/sos-extras-commands.conf` or similar, by default empty)
- interpret each line as a command to execute
- call `add_cmd_output` to them

To collect a file, one would call `cat /path/to/file`. This is a compromise to prevent issues like #1596 when a user wants to collect a file already being collected. At the end, `sos_commands/extras` directory will contain outputs of those commands.

Advantages:
- it will allow much more flexible reaction on *temporarily* enhancing sosreport in collecting some data (until the upstream PR reaches downstream in a while) - this helps downstream maintenance
- the file with the list of extra commands can be easily maintained / content automated / spread over machines, such that all systems one needs to collect some extra data can be easily updated to meet this requirement

Disadvantages:
- no control what is being collected / what command will be executed via `sosreport`. Which means:
  - we still might get an issue similar to #1596 but for commands (hypothetically, now)
  - sosreport could be blamed by failing on some rogue commands a user requested to collect (though it won't be `sosreport`'s mistake)
  - basically the feature opens doors for potential code injection with its security impacts - a user typo/error (or wrong will) would force `sosreport` to execute dangerous/unwanted command (on the other hand: a user who can modify that file can also run the commands directly)
- it will help in rather smaller requirements but wont help when e.g. one needs to add some logic around (iterate over some cmd output to call next commands, run commands under some plugin only etc.)

Any thoughs?

**Comment by TurboTurtle** [comment]
One thought comes to mind - we have a pretty strict rule of "no shell code" in sos plugins. Should we do any inspection of commands listed in the config file? For example, should we check for obvious shell code constructs and refuse to run those?

**Comment by lzap** [comment]
I like the idea, I am worried about different sos cadence too. I had similar idea in mind actually and created patch for our foreman-maintain tool to provide the same: https://github.com/theforeman/foreman_maintain/pull/254 but if this is implemented in sosreport in a generic way, there would be no need of this in Foreman.

**Comment by pmoravec** [comment]
#1628 seems a precondition here. Since my plan is to have `extras` plugin that will list files in directory `/etc/sos.extras.d`, and for each such file, it will execute commands from it.

Supporting multiple files will ease maintenance of them - i.e. if one needs to collect X and Y, it can be automated via puppet/ansible to have file of relevant content - that does not interfere with possible other requirements of extra collection. (though I dont think this will be used often, it is easy to have it implemented).

And for this purposes, it will be great to store individual commands outputs from different files in different subdirectories under `sos_commands/extras/<subdir>`.

**Comment by pmoravec** [comment]
I haven't implemented any check for "no shell code" since that can be hard to properly identify. Anyway, I added there skipping hidden files what prevents uncaught exceptions when having `/etc/sos.plugins.d/somefile` opened in `vim` (that creates its swap file) - sosreport attempts to execute "commands" from the binary swap file, otherwise :)

**Comment by pmoravec** [comment]
Two comments here:

1) There should not be just one file, but one directory. That will allow to deploy multiple independent pieces of data to collect (say, one needs to collect extra data relevant to platform itself, and independently on it, with different "lifecycle", to collect some data for OpenStack or any other layered product). The "extras" plugin will traverse the dir, and for each file, it will collect the data to a specific subdir.

2) I doubt we can implement some check automatically detecting what commands are safe to collect and what aren't. So we would have to rely on users to provide just safe commands. Maybe we shall postulate such requirements as a text and "just" advice users to follow that?
